integration/spark2/src/main/scala/org/apache/spark/util/CarbonClassReflectionUtils.scala

```diff
@@ -0,0 +1,195 @@
+        .newInstance(alias.getOrElse(""), relation, Option(view)).asInstanceOf[SubqueryAlias]
+      subqueryAlias
+    } else if (sparkSession.version.contains("2.2")) {
+      // SubqueryAlias(table.output.map(_.withQualifier(Some(table.tableName))).toString(),
+      //  Project(projList, relation))
+      val clazz = Utils.classForName("org.apache.spark.sql.catalyst.plans.logical.SubqueryAlias")
+      val ctor = clazz.getConstructors.head
+      ctor.setAccessible(true)
+      val subqueryAlias = ctor
+        .newInstance(alias.getOrElse(""), relation).asInstanceOf[SubqueryAlias]
+      subqueryAlias
+    } else {
+      throw new UnsupportedOperationException("Unsupported Spark version")
+    }
+  }
+
+  def getOverWrite[T: TypeTag : reflect.ClassTag](name: String, obj: T): Boolean = {
+    var overwriteboolean: Boolean = false
+    val im = rm.reflect(obj)
+    for (l <- im.symbol.typeSignature.members.filter(_.name.toString.contains("enabled"))) {
+      overwriteboolean = im.reflectField(l.asTerm).get.asInstanceOf[Boolean]
+    }
+    overwriteboolean
+  }
+
+  def getOverWriteOption[T: TypeTag : reflect.ClassTag](name: String, obj: T): Boolean = {
+    var overwriteboolean: Boolean = false
+    val im = rm.reflect(obj)
+    for (m <- typeOf[T].members.filter(!_.isMethod)) {
+      if (m.toString.contains("overwrite")) {
+        val typ = m.typeSignature
+        if (typ.toString.contains("Boolean")) {
+          // Spark2.2
+          overwriteboolean = im.reflectField(m.asTerm).get.asInstanceOf[Boolean]
+        } else {
+          overwriteboolean = getOverWrite("enabled", im.reflectField(m.asTerm).get)
+        }
+      }
+    }
+    overwriteboolean
+  }
+
+  def getFieldOfCatalogTable[T: TypeTag : reflect.ClassTag](name: String, obj: T): Any = {
+    val im = rm.reflect(obj)
+    val sym = im.symbol.typeSignature.member(TermName(name))
+    val tableMeta = im.reflectMethod(sym.asMethod).apply()
+    tableMeta
+  }
+
+  def getAstBuilder(conf: SQLConf,
+      sqlParser: CarbonSpark2SqlParser,
+      sparkSession: SparkSession): AstBuilder = {
+    if (sparkSession.version.contains("2.1")) {
+      val clazz = Utils.classForName("org.apache.spark.sql.hive.CarbonSqlAstBuilder")
+      val ctor = clazz.getConstructors.head
+      ctor.setAccessible(true)
+      val astBuilder = ctor.newInstance(conf, sqlParser).asInstanceOf[AstBuilder]
+      astBuilder
+    } else if (sparkSession.version.contains("2.2")) {
+      val clazz = Utils.classForName("org.apache.spark.sql.hive.CarbonSqlAstBuilder")
+      val ctor = clazz.getConstructors.head
+      ctor.setAccessible(true)
+      val astBuilder = ctor.newInstance(conf, sqlParser).asInstanceOf[AstBuilder]
+      astBuilder
+    } else {
+      throw new UnsupportedOperationException("Spark version not supported")
+    }
+  }
+
+  def getSessionState(sparkContext: SparkContext, carbonSession: CarbonSession): SessionState = {
+    if (sparkContext.version.contains("2.1")) {
```

Comment:
This is not correct, `2.2.1` contains `2.1`. It will give wrong match. Move version matching to a common utility

Reply:
Modified this to euals instead of contains. Also specifying the actual version i.e. 2.1.0 or 2.2.0

Reply:
i think it need to use startswith("2.1"). if use equals, when i use spark 2.2.1, it will be failed